pkg/api/types.go

```diff
@@ -1733,6 +1733,46 @@ type NodeStatus struct {
 	Images []ContainerImage `json:"images,omitempty"`
 }
 
+// AvoidPods desribes pods that should avoid this node. This is the value for a
+// Node annotation with key scheduler.alpha.kubernetes.io/preferAvoidPods and
+// will eventually become a field of NodeStatus.
+type AvoidPods struct {
+	// Bounded-sized list of signatures of pods that should avoid this node, sorted
+	// in timestamp order from oldest to newest. Size of the slice is unspecified.
+	PreferAvoidPods []PreferAvoidPodsEntry `json:"preferAvoidPods,omitempty"`
+}
+
+// Describes a class of pods that should avoid this node.
+type PreferAvoidPodsEntry struct {
+	// The class of pods.
+	PodSignature PodSignature `json:"podSignature"`
+	// Time at which this entry was added to the list.
+	EvictionTime unversioned.Time `json:"evictionTime,omitempty"`
+	// (brief) reason why this entry was added to the list.
+	Reason string `json:"reason,omitempty"`
+	// Human readable message indicating why this entry was added to the list.
+	Message string `json:"message,omitempty"`
+}
+
+// Describes the class of pods that should avoid this node.
+// Exactly one field should be set.
+type PodSignature struct {
+	// Reference to controller whose pods should avoid this node.
+	PodController *PodController `json:"podController,omitempty"`
+}
+
+// PodController contains enough information to let you inspect or modify the referred controller.
+type PodController struct {
+	// Kind of the controller.
+	Kind string `json:"kind"`
+	// Namespace of the controller.
+	Namespace string `json:"namespace"`
+	// Name of the controller.
+	Name string `json:"name"`
+	// API version of the controller.
+	APIVersion string `json:"apiversion"`
```

Comment:
can you describe what this would be set to for a pod created by a controller that is surfaced in multiple versions of the API?